### PR TITLE
feat(airbyte): Allow for special chars in streams

### LINKED ISSUE #11861: Airbyte assets with '/' in stream name produce an illegal asset name
### Dagster version

dagster, version 1.1.13

### What's the issue?

My environment:
* Airtable is a source in Airbyte. The streams from Airtable look like 'my-base-name/table-name'
* Using the dagster-airbyte extension and `load_assets_from_airbyte_instance` to define assets from Airbyte

When the assets for Airtable (via Airbyte) are created, an exception is thrown saying that e.g. 'my_base_name/table_name' is an invalid asset name.

### What did you expect to happen?

The assets should be loaded successfully.

### How to reproduce?

Following the [Airbyte + Dagster doc](https://docs.dagster.io/integrations/airbyte#loading-airbyte-asset-definitions-from-an-airbyte-instance), but with a stream that includes a '/' in its name.

### Deployment type

Local

### Deployment details

_No response_

### Additional information

Though I haven't tested it, I believe dashes in names are okay because Dagster is converting them to underscores in asset_decorator.py

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by chollinger93**
Closes #11861

## tl;dr

Adds an optional `stream_to_asset_map` argument to `build_airbyte_assets` to support the Airbyte prefix setting with special characters, e.g. a `/` for an `s3` or `gcs` target (or, really, any targets that aren't w/in the `^[A-Za-z0-9_]+$` spec).

## Details / Test Case

Set up an Airbyte connection like this:

image

When you define your asset as 

```python
with_resources(
    construct_airbyte_assets(
        connection_id="uud",
        destination_tables="e2e/data"
    ),
    {"airbyte": self._airbyte_resource()},
)
```

You will be greeted with `e2e/data" is not a valid name in Dagster. Names must be in regex ^[A-Za-z0-9_]+$`.

If you run

```python
with_resources(
    construct_airbyte_assets(
        connection_id="uud",
        destination_tables="data"
    ),
    {"airbyte": self._airbyte_resource()},
)
```

You will get `DagsterStepOutputNotFoundError`.

The reason for this is, the result of the API communication to Airbyte via Dagster looks a little something like this:

```json
{
    "job": {
        "id": 22,
        "configType": "sync",
        "configId": "$uuid",
        "enabledStreams": [
            {
                "name": "data"
            }
        ],
        "createdAt": 1692286065,
        "updatedAt": 1692286068,
        "status": "succeeded"
    },
    "attempts": [
        {
            "attempt": {
                "id": 0,
                "status": "succeeded",
                "createdAt": 1692286065,
                "updatedAt": 1692286068,
                "endedAt": 1692286068,
                "bytesSynced": 1392,
                "recordsSynced": 100,
                "totalStats": {
                    "recordsEmitted": 100,
                    "bytesEmitted": 1392,
                    "stateMessagesEmitted": 0,
                    "recordsCommitted": 100
                },
                "streamStats": [
                    {
                        "streamName": "e2e/data",
                        "stats": {
                            "recordsEmitted": 100,
                            "bytesEmitted": 1392,
                            "recordsCommitted": 100
                        }
                    }
                ]
            },
```

`attempts.attempt[n].streamStats.streamname` is used for the output at the end in `utils.generate_materializations`, whereas the `destination_tables` property is used for asset names.

Now, if you run this code version with the new mapping argument:

```python
with_resources(
    construct_airbyte_assets(
        connection_id="uud",
        destination_tables="data",
        stream_to_asset_map={"e2e/data": "data"},
    ),
    {"airbyte": self._airbyte_resource()},
)
```

Everything works, since it'll correctly map the `streamName` to the Dagster `asset`, solving this without any need for `string.replace()` shenanigans. 

image

```bash
python -m pytest python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/*.py 
285 passed, 131 warnings in 1279.47s (0:21:19)
```

### CODE REVIEW

**Comment by smackesey** [comment]
cc @benpankow

**Review by benpankow — CHANGES_REQUESTED** [review]
Looks good to me 👍

**Review by benpankow — APPROVED** [review]
